[jiterator] Reduce templating in jitted_gpu_kernel_impl

peterbell10 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #80138
* __->__ #80103

Previously, a new `jitted_gpu_kernel_impl` was instantiated for every
combination of kernel and data types. This adds a new intermediate,
`jitted_gpu_kernel_generic`, which is only templated on the arity of
the input function. So, the compiler is free to re-use this code
between different kernels. `UnaryOperators.cu` as an example will
only need to compile one version.

This is achieved by:
1. Hoisting static variables out of the `launch_` functions and into
   `JittedKernelVariantCache`, stored in `jitted_gpu_kernel_impl`,
   which is templated on the kernel name and dtypes.
2. Moving arguments describing the kernel's static properties
   (e.g. `name` and `f_inputs_type`) into runtime variables
   which are packaged into a new `jit::KernelDescriptor` struct.
3. changing `extra_args` from a tuple to `c10::ArrayRef<void*>`

We can expect benefits in both binary size and compile times. On my
build, I see an 11 MB reduction in binary size for `libtorch_cuda.so`
and this saving scales linearly with the number of jiterated kernels.

peterbell10 committed:
[jiterator] Reduce templating in jitted_gpu_kernel_impl

Previously, a new `jitted_gpu_kernel_impl` was instantiated for every
combination of kernel and data types. This adds a new intermediate,
`jitted_gpu_kernel_generic`, which is only templated on the arity of
the input function. So, the compiler is free to re-use this code
between different kernels. So `UnaryOperators.cu` as an example will
only need to compile one version.

This is achieved by:
1. Hoisting static variables out of the `launch_` functions and into
   `JittedKernelVariantCache`, stored in `jitted_gpu_kernel_impl`,
   which is templated on the kernel name and dtypes.
2. Moving arguments describing the kernel's static properties
   (e.g. `name` and `f_inputs_type`) into runtime variables
   which are packaged into a new `jit::KernelDescriptor` struct.
3. changing `extra_args` from a tuple to `c10::ArrayRef<void*>`

We can expect benefits in both binary size and compile times. On my
build, I see an 11 MB reduction in binary size for `libtorch_cuda.so`
and this saving scales linearly with the number of jiterated kernels.

[ghstack-poisoned]

peterbell10 committed:
Update on "[jiterator] Reduce templating in jitted_gpu_kernel_impl"


Previously, a new `jitted_gpu_kernel_impl` was instantiated for every
combination of kernel and data types. This adds a new intermediate,
`jitted_gpu_kernel_generic`, which is only templated on the arity of
the input function. So, the compiler is free to re-use this code
between different kernels. So `UnaryOperators.cu` as an example will
only need to compile one version.

This is achieved by:
1. Hoisting static variables out of the `launch_` functions and into
   `JittedKernelVariantCache`, stored in `jitted_gpu_kernel_impl`,
   which is templated on the kernel name and dtypes.
2. Moving arguments describing the kernel's static properties
   (e.g. `name` and `f_inputs_type`) into runtime variables
   which are packaged into a new `jit::KernelDescriptor` struct.
3. changing `extra_args` from a tuple to `c10::ArrayRef<void*>`

We can expect benefits in both binary size and compile times. On my
build, I see an 11 MB reduction in binary size for `libtorch_cuda.so`
and this saving scales linearly with the number of jiterated kernels.

[ghstack-poisoned]

peterbell10 committed:
Update on "[jiterator] Reduce templating in jitted_gpu_kernel_impl"


Previously, a new `jitted_gpu_kernel_impl` was instantiated for every
combination of kernel and data types. This adds a new intermediate,
`jitted_gpu_kernel_generic`, which is only templated on the arity of
the input function. So, the compiler is free to re-use this code
between different kernels. `UnaryOperators.cu` as an example will
only need to compile one version.

This is achieved by:
1. Hoisting static variables out of the `launch_` functions and into
   `JittedKernelVariantCache`, stored in `jitted_gpu_kernel_impl`,
   which is templated on the kernel name and dtypes.
2. Moving arguments describing the kernel's static properties
   (e.g. `name` and `f_inputs_type`) into runtime variables
   which are packaged into a new `jit::KernelDescriptor` struct.
3. changing `extra_args` from a tuple to `c10::ArrayRef<void*>`

We can expect benefits in both binary size and compile times. On my
build, I see an 11 MB reduction in binary size for `libtorch_cuda.so`
and this saving scales linearly with the number of jiterated kernels.

[ghstack-poisoned]

peterbell10 committed:
Update on "[jiterator] Reduce templating in jitted_gpu_kernel_impl"


Previously, a new `jitted_gpu_kernel_impl` was instantiated for every
combination of kernel and data types. This adds a new intermediate,
`jitted_gpu_kernel_generic`, which is only templated on the arity of
the input function. So, the compiler is free to re-use this code
between different kernels. `UnaryOperators.cu` as an example will
only need to compile one version.

This is achieved by:
1. Hoisting static variables out of the `launch_` functions and into
   `JittedKernelVariantCache`, stored in `jitted_gpu_kernel_impl`,
   which is templated on the kernel name and dtypes.
2. Moving arguments describing the kernel's static properties
   (e.g. `name` and `f_inputs_type`) into runtime variables
   which are packaged into a new `jit::KernelDescriptor` struct.
3. changing `extra_args` from a tuple to `c10::ArrayRef<void*>`

We can expect benefits in both binary size and compile times. On my
build, I see an 11 MB reduction in binary size for `libtorch_cuda.so`
and this saving scales linearly with the number of jiterated kernels.

[ghstack-poisoned]

peterbell10 committed:
Update on "[jiterator] Reduce templating in jitted_gpu_kernel_impl"


Previously, a new `jitted_gpu_kernel_impl` was instantiated for every
combination of kernel and data types. This adds a new intermediate,
`jitted_gpu_kernel_generic`, which is only templated on the arity of
the input function. So, the compiler is free to re-use this code
between different kernels. `UnaryOperators.cu` as an example will
only need to compile one version.

This is achieved by:
1. Hoisting static variables out of the `launch_` functions and into
   `JittedKernelVariantCache`, stored in `jitted_gpu_kernel_impl`,
   which is templated on the kernel name and dtypes.
2. Moving arguments describing the kernel's static properties
   (e.g. `name` and `f_inputs_type`) into runtime variables
   which are packaged into a new `jit::KernelDescriptor` struct.
3. changing `extra_args` from a tuple to `c10::ArrayRef<void*>`

We can expect benefits in both binary size and compile times. On my
build, I see an 11 MB reduction in binary size for `libtorch_cuda.so`
and this saving scales linearly with the number of jiterated kernels.

[ghstack-poisoned]

ngimel commented on aten/src/ATen/native/cuda/CUDAJitLoops.cuh:
```diff
@@ -291,18 +231,55 @@ void jitted_gpu_kernel_impl(
+  static const auto desc = at::cuda::jit::make_kernel_descriptor<
+    result_type, f_inputs_type, arity, ExtraArgs...>(name, f);
+
+  auto &cache = device_caches[iter.device().index()];
```
I'm not sure this is correct (and I'm surprised tests don't catch it) for the case where the first argument is cpu scalar. `opmath_jitted_gpu_kernel_with_scalars` correctly sets device guard for this case before calling `jitted_gpu_kernel` and `jitted_gpu_kernel_impl`, but here cache could use other device.

peterbell10 replied:
I wrote this because it's equivalent to what was there before. That said, `iter.device()` is the output tensor's device, so I think this should work correctly.

peterbell10 committed:
Rebase on "[jiterator] Reduce templating in jitted_gpu_kernel_impl"


Previously, a new `jitted_gpu_kernel_impl` was instantiated for every
combination of kernel and data types. This adds a new intermediate,
`jitted_gpu_kernel_generic`, which is only templated on the arity of
the input function. So, the compiler is free to re-use this code
between different kernels. `UnaryOperators.cu` as an example will
only need to compile one version.

This is achieved by:
1. Hoisting static variables out of the `launch_` functions and into
   `JittedKernelVariantCache`, stored in `jitted_gpu_kernel_impl`,
   which is templated on the kernel name and dtypes.
2. Moving arguments describing the kernel's static properties
   (e.g. `name` and `f_inputs_type`) into runtime variables
   which are packaged into a new `jit::KernelDescriptor` struct.
3. changing `extra_args` from a tuple to `c10::ArrayRef<void*>`

We can expect benefits in both binary size and compile times. On my
build, I see an 11 MB reduction in binary size for `libtorch_cuda.so`
and this saving scales linearly with the number of jiterated kernels.

[ghstack-poisoned]

peterbell10 commented:
@pytorchbot merge -g

b0noI commented:
@pytorchmergebot revert this

error:

```
Summary: 
caffe2/aten/src/ATen/native/cuda/jit_utils.h:68:41: error: comparison of integers of different signs: 'size_t' (aka 'unsigned long') and 'int' [-Werror,-Wsign-compare]
  TORCH_INTERNAL_ASSERT(pointers.size() == 1 + desc.nInputs);
                        ~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~
caffe2/aten/src/ATen/native/cuda/jit_utils.h:68:41: error: comparison of integers of different signs: 'size_t' (aka 'unsigned long') and 'int' [-Werror,-Wsign-compare]
  TORCH_INTERNAL_ASSERT(pointers.size() == 1 + desc.nInputs);
                        ~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~
```

b0noI commented:
@pytorchmergebot revert -m "internal breakage" -c nosignal

error:
```
Summary: 
caffe2/aten/src/ATen/native/cuda/jit_utils.h:68:41: error: comparison of integers of different signs: 'size_t' (aka 'unsigned long') and 'int' [-Werror,-Wsign-compare]
  TORCH_INTERNAL_ASSERT(pointers.size() == 1 + desc.nInputs);
                        ~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~
caffe2/aten/src/ATen/native/cuda/jit_utils.h:68:41: error: comparison of integers of different signs: 'size_t' (aka 'unsigned long') and 'int' [-Werror,-Wsign-compare]
  TORCH_INTERNAL_ASSERT(pointers.size() == 1 + desc.nInputs);
                        ~~~~~~~~~~~~~~~ ^  ~~~~~~~~~~~~~~~~
```

b0noI commented:
@pytorchmergebot revert -m "internal breakage" -c nosignal

peterbell10 committed:
Update on "[jiterator] Reduce templating in jitted_gpu_kernel_impl"


Previously, a new `jitted_gpu_kernel_impl` was instantiated for every
combination of kernel and data types. This adds a new intermediate,
`jitted_gpu_kernel_generic`, which is only templated on the arity of
the input function. So, the compiler is free to re-use this code
between different kernels. `UnaryOperators.cu` as an example will
only need to compile one version.

This is achieved by:
1. Hoisting static variables out of the `launch_` functions and into
   `JittedKernelVariantCache`, stored in `jitted_gpu_kernel_impl`,
   which is templated on the kernel name and dtypes.
2. Moving arguments describing the kernel's static properties
   (e.g. `name` and `f_inputs_type`) into runtime variables
   which are packaged into a new `jit::KernelDescriptor` struct.
3. changing `extra_args` from a tuple to `c10::ArrayRef<void*>`

We can expect benefits in both binary size and compile times. On my
build, I see an 11 MB reduction in binary size for `libtorch_cuda.so`
and this saving scales linearly with the number of jiterated kernels.

[ghstack-poisoned]

mehtanirav commented:
@pytorchmergebot revert -m "internal breakage" -c nosignal

```
c:\cygwin\data\sandcastle\boxes\trunk-hg-full-fbsource\xplat\caffe2\aten\src\aten\native\cuda\cudajitloops.cuh(63): error C2100: illegal indirection
```

peterbell10 commented:
`ghstack` was failing so I've opened a new pull request, #80967.

@mehtanirav is there any other context to that error message that could help figure out what's going on? Which translation unit is being compiled for example?
